src/Span.ts

```diff
@@ -503,27 +484,73 @@ const setHttpHeader = (context: TypeSpanContext, type: HttpRequestOptions, heade
 
     case HttpRequestOptions.w3c:
     default: {
-      const version = Buffer.alloc(1).fill(0)
-      const flagsForBuff = (flags && sampled) ? (flags | sampled) : flags ? flags : sampled ? sampled : 0x00
-      const flagsBuffer = Buffer.alloc(1).fill(flagsForBuff)
-      const traceIdBuff = Buffer.from(traceId, 'hex')
-      const spanIdBuff = Buffer.from(spanId, 'hex')
-      const parentSpanIdBuff = parentSpanId && Buffer.from(parentSpanId, 'hex')
-      let result = {}
-      let W3CHeaders = parentSpanIdBuff
-        ? new TraceParent(Buffer.concat([version, traceIdBuff, spanIdBuff, flagsBuffer, parentSpanIdBuff]))
-        : new TraceParent(Buffer.concat([version, traceIdBuff, spanIdBuff, flagsBuffer]))
-      if (headers.tracestate) {
-        return Object.assign({ traceparent: W3CHeaders.toString() }, { tracestate: createW3CTracestate(headers.tracestate, spanId), headers })
-      }
-      return Object.assign({ traceparent: W3CHeaders.toString() }, headers)
+      return {
+        ...headers, 
+        ...{
+          traceparent: context.tags!.traceparent,
+          tracestate: headers.tracestate ? createW3CTracestate(context, headers.tracestate) : context.tags!.tracestate }}
     }
   }
 }
 
+const encodeTracestate = (context: TypeSpanContext): {[key: string]: string} => {
+  return {
+    vendor: Config.EVENT_LOGGER_VENDOR_PREFIX,
+    opaqueValue: context.spanId
+  }
+}
+
+const createW3CTracestate = (spanContext: TypeSpanContext, tracestate?: string): string => {
+  const { vendor, opaqueValue } = encodeTracestate(spanContext)
+  if (!tracestate && Config.EVENT_LOGGER_TRACE_HEADERS_ENABLED) {
+    return `${vendor}=${opaqueValue}`
+  }
+  let tracestateArray = (tracestate!.split(','))
```

Comment:
What if `tracestate` is null or undefined, and `Config.EVENT_LOGGER_TRACE_HEADERS_ENABLED` is falsy? I'm pretty sure this would throw an exception.

Reply:
it will actually be falsy and won't throw, I've tested it https://jsbin.com/pepevayagu/edit?html,js,console,output

Reply:
I mean that there is a fallthrough case where you can get to line `508` and tracestate is `undefined`, which means you are then calling `.split(',')` on an undefined value.

This would happen when `tracestate` hasn't been initialized, and `EVENT_LOGGER_TRACE_HEADERS_ENABLED` is false. Please correct me if I'm wrong...

Reply:
@vgenev I know this is merged, but I'd still like some clarity on the above when you get a chance.